File: plugins/room_scan/android/src/main/kotlin/com/taskfleet/room_scan/bridge/TapPolygonBridge.kt
package com.taskfleet.room_scan.bridge

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

/**
 * Converts a list of tapped XZ floor points (metres) into a
 * FloorPlanScanResult-shaped map suitable for sending across the event
 * channel. Mirrors `Bridge/TapPolygonBridge.swift` on iOS exactly.
 */
object TapPolygonBridge {
    fun toJson(
        polygon: List<FloatArray>,
        captureId: String,
        engine: String
    ): Map<String, Any?> {
        val ccw = ensureCCW(polygon)
        val polyJson = ccw.map { mapOf("x" to it[0].toDouble(), "y" to it[1].toDouble()) }
        val roomId = UUID.randomUUID().toString()
        val room = mapOf(
            "id" to roomId,
            "label" to null,
            "floorPolygonMeters" to polyJson,
            "roomToWorld" to mapOf("x" to 0.0, "y" to 0.0, "yaw" to 0.0)
        )
        return mapOf(
            "captureId" to captureId,
            "engine" to engine,
            "confidence" to 0.5,
            "rooms" to listOf(room),
            "openings" to emptyList<Any>(),
            "objects" to emptyList<Any>(),
            "capturedAt" to iso8601(Date())
        )
    }

    private fun ensureCCW(polygon: List<FloatArray>): List<FloatArray> {
        var sum = 0f
        for (i in polygon.indices) {
            val p = polygon[i]
            val q = polygon[(i + 1) % polygon.size]
            sum += (q[0] - p[0]) * (q[1] + p[1])
        }
        return if (sum > 0) polygon.reversed() else polygon
    }

    private fun iso8601(d: Date): String {
        val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        fmt.timeZone = TimeZone.getTimeZone("UTC")
        return fmt.format(d)
    }
}

File: plugins/room_scan/ios/Classes/Bridge/TapPolygonBridge.swift
import Foundation
import simd

/// Converts a list of tapped XZ floor points (metres) into a
/// FloorPlanScanResult-shaped JSON payload. Used by the manual-tap engines
/// (ARKit on iOS, ARCore on Android via a parallel implementation).
enum TapPolygonBridge {
    static func toJson(
        polygon: [SIMD2<Float>],
        captureId: String,
        engine: String
    ) -> [String: Any] {
        let ccw = ensureCCW(polygon)
        let polyJson: [[String: Any]] = ccw.map { p in
            ["x": Double(p.x), "y": Double(p.y)]
        }
        let roomId = UUID().uuidString
        let room: [String: Any] = [
            "id": roomId,
            "label": NSNull(),
            "floorPolygonMeters": polyJson,
            "roomToWorld": ["x": 0, "y": 0, "yaw": 0],
        ]
        return [
            "captureId": captureId,
            "engine": engine,
            "confidence": 0.5,
            "rooms": [room],
            "openings": [],
            "objects": [],
            "capturedAt": ISO8601DateFormatter().string(from: Date()),
        ]
    }

    private static func ensureCCW(_ polygon: [SIMD2<Float>]) -> [SIMD2<Float>] {
        var sum: Float = 0
        for i in 0..<polygon.count {
            let p = polygon[i]
            let q = polygon[(i + 1) % polygon.count]
            sum += (q.x - p.x) * (q.y + p.y)
        }
        return sum > 0 ? polygon.reversed() : polygon
    }
}
